// structured-headers/dist/serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeItem = exports.serializeDictionary = exports.serializeList = exports.SerializeError = void 0;
const types_1 = require("./types");
const token_1 = require("./token");
const util_1 = require("./util");
class SerializeError extends Error {
}
exports.SerializeError = SerializeError;
function serializeList(input) {
    return input.map(value => {
        if (util_1.isInnerList(value)) {
            return serializeInnerList(value);
        }
        else {
            return serializeItem(value);
        }
    }).join(', ');
}
exports.serializeList = serializeList;
function serializeDictionary(input) {
    return Array.from(input.entries()).map(([key, value]) => {
        let out = serializeKey(key);
        if (value[0] === true) {
            out += serializeParameters(value[1]);
        }
        else {
            out += '=';
            if (util_1.isInnerList(value)) {
                out += serializeInnerList(value);
            }
            else {
                out += serializeItem(value);
            }
        }
        return out;
    }).join(', ');
}
exports.serializeDictionary = serializeDictionary;
function serializeItem(input) {
    return serializeBareItem(input[0]) + serializeParameters(input[1]);
}
exports.serializeItem = serializeItem;
function serializeInnerList(input) {
    return `(${input[0].map(value => serializeItem(value)).join(' ')})${serializeParameters(input[1])}`;
}
function serializeBareItem(input) {
    if (typeof input === 'number') {
        if (Number.isInteger(input)) {
            return serializeInteger(input);
        }
        return serializeDecimal(input);
    }
    if (typeof input === 'string') {
        return serializeString(input);
    }
    if (input instanceof token_1.Token) {
        return serializeToken(input);
    }
    if (input instanceof types_1.ByteSequence) {
        return serializeByteSequence(input);
    }
    if (typeof input === 'boolean') {
        return serial
